fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,943 @@
+/// could be mapped to
+/// 100..104 -> 0..4
+/// 50000..50001 -> 5..6
+///
+/// Compact space 0..=6 requires much less bits than 100..=50001
+///
+/// The codec is created to compress ip addresses, but may be employed in other use cases.
+use std::{
+    cmp::Ordering,
+    collections::BinaryHeap,
+    convert::{TryFrom, TryInto},
+    io::{self, Write},
+    net::{IpAddr, Ipv6Addr},
+    ops::RangeInclusive,
+};
+
+use common::{BinarySerializable, CountingWriter, VIntU128};
+use ownedbytes::OwnedBytes;
+use tantivy_bitpacker::{self, BitPacker, BitUnpacker};
+
+use crate::column::{ColumnV2, ColumnV2Ext};
+
+pub fn ip_to_u128(ip_addr: IpAddr) -> u128 {
+    let ip_addr_v6: Ipv6Addr = match ip_addr {
+        IpAddr::V4(v4) => v4.to_ipv6_mapped(),
+        IpAddr::V6(v6) => v6,
+    };
+    u128::from_be_bytes(ip_addr_v6.octets())
+}
+
+/// The cost per blank is quite hard actually, since blanks are delta encoded, the actual cost of
+/// blanks depends on the number of blanks.
+///
+/// The number is taken by looking at a real dataset. It is optimized for larger datasets.
+const COST_PER_BLANK_IN_BITS: usize = 36;
+
+/// The range of a blank in value space.
+///
+/// A blank is an unoccupied space in the data.
+/// Ordered by size
+///
+/// Use the try_into(), invalid ranges will be rejected.
+///
+/// TODO: move to own module to force try_into
+#[derive(Debug, Eq, PartialEq, Clone)]
+struct BlankRange {
+    blank_range: RangeInclusive<u128>,
+}
+impl TryFrom<RangeInclusive<u128>> for BlankRange {
+    type Error = &'static str;
+    fn try_from(range: RangeInclusive<u128>) -> Result<Self, Self::Error> {
+        let blank_size = range.end().saturating_sub(*range.start());
+        if blank_size < 2 {
+            Err("invalid range")
+        } else {
+            Ok(BlankRange { blank_range: range })
+        }
+    }
+}
+impl BlankRange {
+    fn blank_size(&self) -> u128 {
+        self.blank_range.end() - self.blank_range.start()
+    }
+}
+
+impl Ord for BlankRange {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.blank_size().cmp(&other.blank_size())
+    }
+}
+impl PartialOrd for BlankRange {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.blank_size().partial_cmp(&other.blank_size())
+    }
+}
+
+/// Put the blanks for the sorted values into a binary heap
+fn get_blanks(values_sorted: &[u128]) -> BinaryHeap<BlankRange> {
+    let mut blanks: BinaryHeap<BlankRange> = BinaryHeap::new();
+    let mut add_range = |blank_range: RangeInclusive<u128>| {
+        let blank_range: Result<BlankRange, _> = blank_range.try_into();
+        if let Ok(blank_range) = blank_range {
+            blanks.push(blank_range);
+        }
+    };
+    for values in values_sorted.windows(2) {
+        let blank_range = values[0] + 1..=values[1] - 1;
+        add_range(blank_range);
+    }
+    if let Some(first_val) = values_sorted.first().filter(|first_val| **first_val != 0) {
```

Comment:
nitpick... 

The use of filter here is correct but makes it difficult to read.

```
if let Some(first_val) = values_sorted.first() {
    if first_val > 0 {
       // ...
    }
}
```

would be easier.